### ISSUE TITLE
[rl] Add a skip counter to AgentStats for FiniteLossGuard skips (ADR 0056 remainder)

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#346]
## Summary

Add an `is_finite()` guard on the loss before `backward()` in each agent's learn step, skipping the update and counting the skip. Issue #173 deferred this on cost grounds. **That reasoning was wrong**, and the correction is the point of this issue.

## The premise #173 deferred on is false

#173 argued:

> "Skip-and-warn on non-finite loss" requires reading the loss scalar host-side (`into_scalar`) every update, forcing a **device→host sync per step** on wgpu. That is a real throughput cost and a design tradeoff needing its own discussion, not a blanket helper.

**The read already happens**, unconditionally, immediately before `backward()`, in every agent:

- `dqn_agent.rs:501` — `let loss_value = loss_tensor.clone().into_scalar().elem::<f32>();` then `:503` `let grads = loss_tensor.backward();`
- `sac_agent.rs:615-616` — both critic losses read host-side; `backward()` at `:619` and `:627`
- `td3_agent.rs:587-588`, `c51_agent.rs:570` — same shape
- `ppo_agent.rs:537` — under a comment that literally reads `// Diagnostics before backward (fresh copies).`

The scalar is already on the host, already in a local, already positioned before the gradient computation — because every agent reports loss in its stats and cannot drop the read anyway. The guard is:

```rust
if !loss_value.is_finite() { /* count, skip */ }
```

an `f32` predicate on an existing host value. **Zero additional syncs.**

## The other obstacle is also gone

#173 correctly noted that ADR 0046 / #167 removed the *other* blocker: under the old `Option::take()` idiom an early `return` left the field `None` permanently, bricking the agent exactly as a panic would. ADR 0046's `Slot<M>` reorders to forward-before-take, so the loss exists while the module is only borrowed and an early return leaves the `Slot` intact. (This half of #173's reasoning checks out — verified against ADR 0046 and `algorithms/shared.rs:120`.)

## So what is actually open: semantics, not cost

The real objection is the one ADR 0049 raised against a bare clamp and answered with mandatory telemetry: **silently skipping updates converts a loud divergence into a slow one.** A run that quietly discards 40% of its updates looks like a run that simply learns badly.

Any guard must therefore ship with a skip counter surfaced in the agent's stats, for the same reason `min_log_std` shipped alongside the `log_std` clamp. Design questions worth settling before implementing:

- Skip counter in each agent's stats struct, and a one-shot `tracing::warn!` on first skip.
- Threshold behavior: is a sustained skip rate an error, or just reported?
- Placement: between the existing `into_scalar` and `backward()` in each of the seven learn steps.

## Sequencing

Sequence against **#184** (SAC alpha Adam poisoning), which is an actual *located* NaN origin rather than a generic guard. A guard here masks symptoms; #184 fixes a cause. Worth noting #184's own zero-sync-cost claim is likewise correct — `sac_agent.rs:669` reads `log_prob_mean` host-side outside the autotune block, so its value is materialized regardless.

## Provenance

This absorbs the generic "no `is_finite` guard before `backward()`" rows that the 30-06-2026 code reviews had filed against #173 across SAC (§5.4, §1.3), DQN (§1.3), DDPG (§1.2), and QR-DQN (§1.3). Those rows have been re-pointed here; none of them named a NaN-producing mechanism, which is why they belong in one deliberate design proposal rather than six bug reports.

Related: #173, #184, #192, #167 / ADR 0046, ADR 0049.

**Comment by anthonytorlucci** [comment]
## DDPG confirmation (re-triage of `rl-ddpg.md` §1.2, 2026-07-20)

Confirming the DDPG row this issue already absorbed is still live against the current tree, with current line numbers: `ddpg_agent.rs:519` (`q_mean`), `:522` (`critic_loss`), `:539` (`actor_loss_value`) all extract `.into_scalar().elem::<f32>()` with no finiteness check before being returned in `LearnOutcome` / recorded into `AgentStats`. No behavior change here since the design proposal (skip-and-count with a stats counter) is still pending; just verifying the DDPG instance of the gap hasn't drifted or been separately patched.

Also worth noting for scope: `train.rs`'s own §1.1 (a training-loop-level `is_finite()` check on `LearnOutcome.critic_loss`/`q_mean` that aborts the run) is a different location than this issue's agent-level pre-`backward()` guard, but the same generic gap — flagging here rather than filing separately, since the design questions this issue is already working through (skip counter, threshold, debug-gating) apply equally to a loop-level check.

**Comment by anthonytorlucci** [comment]
File-specific detail from `crates/rlevo-reinforcement-learning/src/algorithms/ppo/policies/categorical.rs` (2026-06-30 PPO categorical-head review, §1.1/§5.3):

`sample_with_logprob` (categorical.rs:170-171) and `evaluate` (categorical.rs:189-190) both compute entropy as `(probs * log_probs).sum_dim(1).neg()` where `probs = log_probs.exp()`. For finite logits this is safe (Burn's `log_softmax` max-subtracts internally, so `0.0 * finite = 0.0`). But if a logit is already `+inf`/`NaN` upstream, `log_softmax` emits `-inf` and `0.0 * -inf = NaN` — this is an **amplifier**, not an origin: it only produces NaN if a non-finite logit already arrived from the network, consistent with this issue's finding that none of the reviewed sites name a NaN-*producing* mechanism. A per-agent finiteness guard before `backward()` (this issue's scope) would catch it at the PPO `update` boundary the same way it does for the other six agents; no categorical-specific fix is needed.

**Comment by anthonytorlucci** [comment]
ADR 0056 (landed for #318) builds the **skip half** of this proposal at every site — 17 loss sites across all 8 agents — and it is correct where it matters most: the skip re-fires on *every* non-finite occurrence and only the `tracing::warn!` is one-shot-latched (`shared.rs:455-477`; test `finite_loss_guard_skip_refires_but_warning_latches_once` at `shared.rs:1337`). Verified gating **both** `backward()` and the optimizer `step_with` at, e.g., `ppo_agent.rs:579`, `dqn_agent.rs:525`, `sac_agent.rs:646/657/711`, `td3_agent.rs:631/642/674`. The device→host-sync objection this issue set out to refute is indeed gone — the guard rides the loss read each agent already did.

It deliberately does **not** build the **counting half** this issue calls a "must" (*"Any guard must therefore ship with a skip counter surfaced in the agent's stats"*). No `AgentStats` field was added — `metrics.rs` is untouched by the diff, and a crate-wide search for `skip_count`/`nan_count`/`non_finite_count` finds nothing. ADR 0056 §3 substitutes a one-shot `warn!`, citing #184's precedent. That is a real remainder, not an oversight: after the first log line, a run discarding 1% of updates and one discarding 40% are indistinguishable, which is precisely the "loud divergence quietly becomes a slow one" failure this issue argued against.

The skip/warn mechanism itself is done and should **not** be reopened for redesign. Retitling this issue to scope it to the genuine remainder — a skip counter surfaced in `AgentStats` — so it isn't re-implemented from scratch. Note the counter touches the metrics/record schema (ADR 0014 territory), which is why it was deliberately kept out of #318's bounded fix and left for a decision here.
